### ISSUE TITLE
Any way around long initial build times?

### ISSUE THREAD

**Issue by mohanzhang** [pbrisbin/docker-heroku-haskell-stack#1]
(I must caveat this question with the fact that my understanding of docker is essentially limited to what I've been able to glean by staring at their cute whale logo... i.e. not much)

I've successfully deployed several times now with `heroku docker:release` and I must say that on the whole, this experience has been a lot better than haskell-on-heroku. I'm not sure what to make of the initial build time though---similarly when the dependencies change and the build has to start over. The build on my machine takes a very long time since it has to build all the dependencies from scratch in these situations (~100 packages).

What do people typically do in this situation? Do they use some kind of on-demand docker instance on AWS to get a bunch of compute power in small bursts? Or is there a better way to structure my application so that it doesn't have to rebuild (esp since I know ahead of time which packages and versions I am using by virtue of building with stack)?

**Comment by pbrisbin** [comment]
It's a good question. So far, I've just been eating the time. It happens infrequently, and costs 30-40 minutes in my case(s) when it does, so it hasn't reached a stage of _unbearability_ for me yet.

There is a way around it, but it means not using this (or any) generalized base image. Basically, you would pre-install large dependencies (yesod, etc) in their own layer, before you `COPY` in your cabal file.

```
# this step takes forever but hopefully you never lose that cache
RUN stack install yesod yesod-form ...

COPY *.cabal ...

# this cache will blow if *.cabal is touched, but it only builds dependencies
# not installed above, so should be quick
RUN stack build --dependencies-only
```

Unfortunately, a generalized base image (like this one) can't really make that assumption (that you'll need yesod, etc) and would have to employ some sort of configuration trick to allow it while remaining generalized -- such a trick may exist, but I haven't explored it yet.

**Comment by mohanzhang** [comment]
Well, I was hoping I could get some more traction on the path I went down before presenting my findings here, but I am very close to getting something working via CircleCI and then deploying it to heroku to boot. I am just "one docker error" away, which seems to be a fairly common feeling when you're using docker, I am learning. It would appear that CircleCI uses its own version of docker. Despite being a paying customer, I've had sluggish response from their support team, so I can't go on record here recommending this avenue to others just yet.

In short, my hypothesis was that CircleCI's machines would be faster than mine---and in the case where the build times were equal, I could at least spare the extra kelvins coming out of my laptop. So far, I'm saving just a couple of minutes on my build times, but it's nice to decouple the build and deploy from my laptop (and more importantly, my laptop fan).
